### PR TITLE
Add NetworkDB docs

### PR DESCRIPTION
**Pull request by talex5**
This documentation addition is based on reading the code in the `networkdb` directory.

### CODE REVIEW

**Comment by talex5** [comment]
Had a go at triggering the UDP size limit problem. It's actually quite tricky to make it send an oversized packet because we have compression turned on, so you need to go quite a long way over the limit before the compressed version is too big.

With compression turned off, I was able to get it to send a 1501 UDP packet when the limit was 1400 (the default), with this hacked-up unit test:

```
func TestNetworkDBSimple(t *testing.T) {
	nodes := 10
	dbs := createNetworkDBInstances(t, nodes, "node", DefaultConfig())
	log.Printf("Created %d DBs", len(dbs))
	wg := sync.WaitGroup{}
	wg.Add(nodes)
	for i := 0; i < nodes; i++ {
		go func(i int) {
			log.Printf("Init %d", i)
			err := dbs[i].sendNodeEvent(NodeEventTypeJoin)
			if err != nil { panic(err) }
			go func() { dbs[i].sendNodeEvent(NodeEventTypeJoin) } ()
			go func() { dbs[i].sendNodeEvent(NodeEventTypeLeave) } ()
			go func() { dbs[i].sendNodeEvent(NodeEventTypeJoin) } ()
			err = dbs[i].JoinNetwork("network-a" + strconv.Itoa(i))
			if err != nil { panic(err) }
			go func() { dbs[i].JoinNetwork("network-b" + strconv.Itoa(i)) } ()
			go func() { dbs[i].JoinNetwork("network-c" + strconv.Itoa(i)) } ()
			go func() { dbs[i].JoinNetwork("network-d" + strconv.Itoa(i)) } ()
			go func() { dbs[i].JoinNetwork("network-e" + strconv.Itoa(i)) } ()
			log.Printf("Done!")
			wg.Done()
		}(i)
	}
	log.Printf("Waiting for %d nodes", nodes)
	wg.Wait()
	log.Printf("Sleeping...")
	time.Sleep(10 * time.Second)
	log.Printf("Shutting down...")
	closeNetworkDBInstances(dbs)
}
```

File: docs/networkdb.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+While a node is active in the cluster, it can join and leave networks.
+When a node wants to join a network, it will send a `NetworkEventTypeJoin` message over the gossip network to the whole cluster.
+It will also perform a bulk-sync of the network-specific state (the tables) with every other node on the network being joined.
+This will allow it to get all the network-specific information quickly.
+The tables will mostly be kept up-to-date by UDP gossip messages between the nodes on that network, but
+each node in the network will also do a full TCP sync of the tables with another random node on the same network from time to time.
+
+When a node wishes to leave a network, it will send a `NetworkEventTypeLeave` via gossip. It will then delete the network's table data.
+When a node hears that another node is leaving a network, it deletes all table entries belonging to the leaving node.
+Deleting an entry in this case means marking it for deletion for a while (so that the deletion can propagate via gossip too).
```
Comment (fcrisciani) [inline docs/networkdb.md]: and more important to protect itself from receiving an old CREATE with lower version and accept it

File: docs/networkdb.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+When a node wishes to leave the cluster, it will send a `NodeEventTypeLeave` message via gossip.
+Nodes receiving this will mark the node as "left".
+The node will then send a memberlist leave message too.
```
Comment (fcrisciani) [inline docs/networkdb.md]: meaning will forward it to others?

File: networkdb/delegate.go [hunk 3]
```diff
@@ -402,6 +403,7 @@ func (d *delegate) NotifyMsg(buf []byte) {
 	d.nDB.handleMessage(buf, false)
 }
 
+// XXX: should this limit be shared?
 func (d *delegate) GetBroadcasts(overhead, limit int) [][]byte {
```
Comment (fcrisciani) [inline networkdb/delegate.go]: Not sure why this PR(https://github.com/docker/libnetwork/pull/1446) changed it this way. To me makes more sense the original code with the simple return...

File: networkdb/cluster.go [hunk 4]
```diff
@@ -492,7 +492,7 @@ func (nDB *NetworkDB) gossip() {
 			nDB.RUnlock()
 
 			if mnode == nil {
-				break
+				break // XXX: shouldn't this be "continue"?
```
Comment (fcrisciani) [inline networkdb/cluster.go]: yep a continue makes more sense

**Pull request merged**